Introduce SymbolObjectCodeRequest

hamishknight opened the pull request:
Add `SymbolObjectCodeRequest`, which emits object code for a specific set of symbols by querying TBDGen for the source of the symbols, and asking SILGen (and eventually IRGen) to emit them.

hamishknight commented:
@swift-ci please test

CodaFi commented on lib/IRGen/IRGen.cpp:
```diff
@@ -79,6 +81,7 @@
 #include "llvm/Transforms/Instrumentation/SanitizerCoverage.h"
 #include "llvm/Transforms/Instrumentation/ThreadSanitizer.h"
 #include "llvm/Transforms/ObjCARC.h"
+#include "llvm/Transforms/Utils/Cloning.h"
```
Curious what motivates this header dependency.

hamishknight replied:
Ah that's a holdover from https://github.com/apple/swift/pull/32964 when we cloned the module with the requested symbols – good spot!

hamishknight committed:
Rename getFiles to getFilesToEmit

Rename the member on ASTLoweringDescriptor and
IRGenDescriptor to make it more explicit it returns
the files that need emitting, rather than just the
files that happen to be present. This distinction
will become important once we start emitting code
only for a specific set of symbols.

hamishknight committed:
[SILGen] Allow lazy type-checking

Remove the type-checking call from
OptimizedIRRequest, and sink it down into SILGen
and IRGen when we come to emit the source files.

hamishknight committed:
Introduce SymbolObjectCodeRequest

Add SymbolObjectCodeRequest, which emits object
code for a specific set of symbols by querying
TBDGen for the source of the symbols, and asking
SILGen (and eventually IRGen) to emit them.

hamishknight committed:
Allow SILGen to emit only a specific set of SILDeclRefs

Extend ASTLoweringDescriptor to store a set of
SILDeclRefs to emit through `emitFunctionDefinition`.

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please smoke test
